standardized-output-composition.md

```diff
@@ -0,0 +1,291 @@
+
+- **Pass/Fail** - A boolean which simply indicates whether the schema was validated.
+- **Basic** - Provides validation information in a flat list structure.
+- **Detailed** - Provides validation information in a condensed hierarchical structure based on the structure of the schema.
+- **Verbose** - Provides validation information in an uncondensed hierarchical structure that matches the exact structure of the schema.
+
+Implementations SHOULD specify which levels of validation they support.
+
+## Minimum Information
+
+Beyond a simplistic pass/fail, additional information is useful to aid in debugging a schema or instance.  Each sub-result SHOULD contain the information contained within this section at a minimum.
+
+A single object that contains all of these components is considered an output unit.
+
+Implementations MAY elect to provide additional information.
+
+### Keyword relative location
+
+The relative location of the validating keyword that follows the validation path.  The value MUST be expressed as a JSON Pointer, and it MUST include `$ref` or `$recursiveRef` segments for any reference jumps **<span style="color:red">(wording?)</span>**.
+
+```
+#/properties/minLength/$ref/minimum
+```
+
+The JSON key for this information is `keywordLocation`.
+
+### Keyword absolute location
+
+The absolute, dereferenced location of the validating keyword.  The value MUST be expressed as an absolute URI, and it MUST NOT include `$ref` or `$recursiveRef` segments.
+
+```
+http://json-schema.org/draft-08/schema#/$defs/nonNegativeInteger/minimum
+```
+
+This information MAY be omitted if either the relative location contains no references or if the schema does not declare an absolute URI as its `$id`.
+
+The JSON key for this information is `absoluteKeywordLocation`.
+
+### Instance location
+
+The location of the JSON value within the instance being validated.  The value MUST be expressed as a JSON Pointer.
+
+The JSON key for this information is `instanceLocation`.
+
+### Error/Annotation
+
+The error or annotation that is produced by the validation.
+
+For errors, the specific wording for the message is not defined by this specification.  Implementations will need to provide this.
+
+The JSON key for failed validations is `error`; for successful validations it is `annotation`.
+
+### Nested results
+
+For the two hierarchical structures, this property will hold nested errors and annotations.
+
+The JSON key for nested results in failed validations is `errors`; for successful validations it is `annotations`.
+
+## Output Structure
+
+The output MUST be an object containing two properties
+
+- `valid` - a boolean value indicating the overall validation success or failure
+- `errors` - the collection of errors or annotations produced by a failed validation
+- `annotations` - the collection of errors or annotations produced by a successful validation
```

Comment:
Only the verbose output would do that.  The primary cases to consider are the applicator keywords (especially `not`). For example, an `allOf` could contain 1 failed subschema, yielding an overall failure, but the verbose output would list *all* of the results. This would need to be inside `errors` since it failed.

I agree it's a little counterintuitive, but I think it works better for automated processing of all the results are together rather than having both `errors` *and* `annotations`.

Reply:
OK, so my follow up question here is:
If a subschema of `not` fails, obviously the `not` means the parent schema asserts valid is true. My question is then what happens to the annotations in the subschema? They only trickle up for schemas for which validation is a positive assertion, no? /ccing @handrews on this one =]

Reply:
Correct, you only return annotations if the assertion result is valid.  Otherwise they are dropped from the final result set (although showing what they were before they were dropped in the more complex output format is fine and useful.